File: chinesenewyear/submissions/accepted/jannik.java
import java.util.*;
public class jannik {
  static class IntList extends ArrayList<Integer> {}

  static boolean[] deleted;
  static IntList[] adj;
  static IntList[] adji;
  static boolean drop(int i) {
    while (!adji[i].isEmpty() && deleted[adji[i].get(adji[i].size() - 1)]) {
      adj[i].remove(adj[i].size() - 1);
      adji[i].remove(adji[i].size() - 1);
    }
    return adj[i].isEmpty();
  }
  static IntList step;
  static int m;
  static void dfs(int i) {
    while (!drop(i)) {
      int j = adj[i].get(adj[i].size() - 1);
      int e = adji[i].get(adji[i].size() - 1);
      deleted[e] = true;
      dfs(j);
      if (e >= m) step.add(-1);
    }
    step.add(i);
  }
  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    int n = in.nextInt();
    m = in.nextInt();
    for (int i = 0; i < 2 * n; i++) in.next();
    adj = new IntList[n];
    adji = new IntList[n];
    for (int i = 0; i < n; i++) {
      adj[i] = new IntList();
      adji[i] = new IntList();
    }
    for (int i = 0; i < m; i++) {
      int a = in.nextInt() - 1, b = in.nextInt() - 1;
      adj[a].add(b);
      adji[a].add(i);
      adj[b].add(a);
      adji[b].add(i);
    }

    int m2 = m;
    for (int pre = -1, i = 0; i < n; i++) {
      if (adj[i].size() % 2 == 0)
        continue;
      if (pre < 0) pre = i;
      else {
        adj[pre].add(i);
        adji[pre].add(m2);
        adj[i].add(pre);
        adji[i].add(m2);
        m2++;
        pre = -1;
      }
    }

    deleted = new boolean[m2];

    ArrayList<List<Integer>> res = new ArrayList<>();
    for (int i = 0; i < n; i++) {
      if (drop(i))
        continue;
      step = new IntList();
      dfs(i);
      int idx = step.indexOf(-1);
      if (idx != -1) {
        step.remove(step.size() - 1);
        Collections.rotate(step, step.size() - idx);
      }
      for (int it = 0, fst = 0; ; it++) {
        if (it == step.size() || step.get(it) < 0) {
          if (fst != it) res.add(step.subList(fst, it));
          if (it == step.size()) break;
          else fst = ++it;
        }
      }
    }
    System.out.println(res.size());
    for (List<Integer> v : res) {
      System.out.println(v.size());
      for (int x : v)
        System.out.print((x + 1) + " ");
      System.out.println();
    }
  }
}

File: chinesenewyear/submissions/accepted/jannik.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

#define sz(a) ((int)(a).size())
#define all(a) (a).begin(),(a).end()

int main() {
	cin.tie(nullptr), ios::sync_with_stdio(false);
	int n, m;
	cin >> n >> m;
	for (int t, i = 0; i < 2 * n; i++) cin >> t;
	vector<vector<pair<int, int>>> adj(n);
	for (int i = 0; i < m; i++) {
		int a, b;
		cin >> a >> b;
		a--, b--;
		adj[a].emplace_back(b, i), adj[b].emplace_back(a, i);
	}

	int m2 = m;
	for (int pre = -1, i = 0; i < n; i++) {
		if (sz(adj[i]) % 2 == 0)
			continue;
		if (pre < 0) pre = i;
		else {
			adj[pre].emplace_back(i, m2);
			adj[i].emplace_back(pre, m2);
			m2++, pre = -1;
		}
	}

	vector<bool> remaining(m2, true);
	const auto drop = [&](int i) {
		while (not adj[i].empty() and not remaining[adj[i].back().second])
			adj[i].pop_back();
		return adj[i].empty();
	};
	vector<int> step;
	const auto dfs = [&](auto&& self, int i) -> void {
		while (not drop(i)) {
			auto[j,e] = adj[i].back();
			remaining[e] = false;
			self(self, j);
			if (e >= m) step.emplace_back(-1);
		}
		step.emplace_back(i);
	};
	vector<vector<int>> res;
	for (int i = 0; i < n; i++) {
		if (drop(i))
			continue;
		step.clear();
		dfs(dfs, i);
		if (auto it = find(all(step), -1); it != step.end()) {
			step.pop_back();
			rotate(step.begin(), it, step.end());
		}
		for (auto it = step.begin(), fst = it; ; it++) {
			if (it == step.end() or *it < 0) {
				if (fst != it) res.emplace_back(fst, it);
				if (it == step.end()) break;
				else fst = ++it;
			}
		}
	}
	cout << sz(res) << '\n';
	for (auto& v : res) {
		cout << sz(v) << '\n';
		for (auto x : v)
			cout << x+1 << ' ';
		cout << '\n';
	}
}
